securedrop_client/app.py

```diff
@@ -147,6 +147,9 @@ def start_app(args, qt_args) -> None:
     configure_logging(args.sdc_home)
     logging.info('Starting SecureDrop Client {}'.format(__version__))
 
+    # init data singleton
+    Data(args.sdc_home)
```

Comment:
I strongly think we should avoid singleton / global patterns as this will complicate testing. I would much rather have us use injection from the top level for this.

Reply:
I'd also rather use injection, but it's a little complicated. To enable accessor patterns like on sqlalchemy objects, like `submission.content` where `.content` points to the content of the downloaded submission (which is held in `<home>/data/<submission.filename>`, not in the database), the object instantiated by sqlalchemy needs some knowledge about the runtime state of the system: namely, the value of `home`. By knowing `home`, the object knows where to look for downloaded files, and can find the downloaded content of itself.

The `data` singleton is just a further abstraction of that idea, of course- it's essentially a small wrapper around a single piece of global state: the value of `home`.

Since we're not instantiating our own objects in these cases (ie, the `Submission` object is instantiated by sqlalchemy, not by our own code), we don't really have the option to inject that bit of runtime state at object-creation time. 

To avoid a globals/singleton, we can manually set that runtime state via `<instance>.data = <data object>` (or similar) in all cases where we want to set or access `.content` in message, reply, or file objects. Since we're only using `.content` in one place so far, that would be pretty easy- I'll give it a shot.

Reply:
Ooh we might be able to be clever with sqlalchemy's "ORM Events" https://docs.sqlalchemy.org/en/latest/orm/events.html. It's an exciting morning over here.

Reply:
My gut says that FS thnigs should be handled by an injected `Storage` class and you pass it a filename or something and it looks it up. It's not as simple as the accessor pattern, but it would de-spaghetti the code some.

Reply:
Yup, my gut says that, too! The hard part is that content is accessed (via accessors) in the GUI classes, which shouldn't have to know anything about the storage layer. So you have to get the content "into" the objects before they're accessed in the GUI code... so to a certain extent, we're stuck with the accessor pattern, like it or not. Unless we break our data abstractions and let the GUI call `Storage.get(submission.filename))` we're sorta stuck, as far as I can tell.

Reply:
Hey folks... this is a really interesting discussion. :-)

I'm with @heartsucker in outlook about singletons.

Also, why the long-running thread rather than using a QTimer based fire-and-forget sync call (i.e. just re-using the code we already have)..? This "feels" (note the quotes) simpler to me, and I'm anxious to retain simplicity and re-use what we already have.

I'd strongly recommend keeping the logic and UI separate. (For instance, the logic layer uses a QTimer to refresh on new message state and call the UI layer if there's a change.)

Also, I'm painfully aware that the above comments are based on my limited understanding of the aim of the ticket... it is *very* likely that I have the wrong end of the stick. ;-)

Reply:
Cool. The singleton's gone in favor of the sqlalchemy ORM Event approach! Hooray. 

The long-running (which is really an always-running, separate thread) has to do with my preference to separate concerns, and for react/redux/re-frame -style functional reactive programming (particularly, triggering actions based on changes in application state, not as a side-effect of other actions). In our current data model we can't really subscribe to DB changes, so this simulates it by periodically polling the DB and looking for changes that the downloader would be interested in: the existence of not-downloaded submissions. 

In this case, if we can guarantee that at any time there will be not more than one of those threads running, then there's not much practical difference between the approach that's implemented here vs periodically running a fire-and-forget thread which would do the exact same thing (examine the database for not-download submissions, then download them), aside from the decoupling of concerns: as-implemented, the "main" thread has nothing to do with coordinating message download. 

That patterns can also extend to Reply and File downloads: unlike `get_remote_data`, which (absent a two-way channel to the server) must be triggered by a client-internal event (a timer expiring, the user clicking the refresh button), the various download actions can _all_ be independent and reactive to the state of the database changing, and do not have to subscribe to events in the main client thread. In my eyes, that's conceptually simpler than asking the main thread to coordinate download threads (aside from their initial spawning at client init time).

All that said, I agree that our priority should be clarity and simplicity for all involved, so I'd be happy to port this to a QTimer-based approach (or port other actions to this one!)